lib/spack/spack/util/windows_registry.py

```diff
@@ -0,0 +1,321 @@
+"""
+
+import os
+import sys
+from contextlib import contextmanager
+
+from llnl.util import tty
+
+is_windows = sys.platform == "win32"
+if is_windows:
+    import winreg
+
+
+class RegistryValue(object):
+    """
+    Class defining a Windows registry entry
+    """
+
+    def __init__(self, name, value, parent_key):
+        self.path = name
+        self.value = value
+        self.key = parent_key
+
+
+class RegistryKey(object):
+    """
+    Class wrapping a Windows registry key
+    """
+
+    def __init__(self, name, handle):
+        self.path = name
+        self.name = os.path.split(name)[-1]
+        self.handle = handle
+        self._keys = []
+        self._values = {}
+
+    @property
+    def values(self):
+        """Returns all subvalues of this key as RegistryValue objects in dictionary
+        of value name : RegistryValue object
+        """
+        self._gather_value_info()
+        return self._values
+
+    @property
+    def subkeys(self):
+        """Returns list of all subkeys of this key as RegistryKey objects"""
+        self._gather_subkey_info()
+        return self._keys
+
+    @property
+    def hkey(self):
+        return self.handle
+
+    def __str__(self):
+        return self.name
+
+    def _gather_subkey_info(self):
+        """Composes all subkeys into a list for access"""
+        if self._keys:
+            return
+        sub_keys, _, _ = winreg.QueryInfoKey(self.handle)
+        for i in range(sub_keys):
+            sub_name = winreg.EnumKey(self.handle, i)
+            sub_handle = winreg.OpenKeyEx(self.handle, sub_name, access=winreg.KEY_READ)
+            self._keys.append(RegistryKey(os.path.join(self.path, sub_name), sub_handle))
+
+    def _gather_value_info(self):
+        """Compose all values for this key into a dict of form value name: RegistryValue Object"""
+        if self._values:
+            return
+        _, values, _ = winreg.QueryInfoKey(self.handle)
+        for i in range(values):
+            value_name, value_data, _ = winreg.EnumValue(self.handle, i)
+            self._values[value_name] = RegistryValue(value_name, value_data, self.handle)
+
+    def get_subkey(self, sub_key):
+        """Returns subkey of name sub_key in a RegistryKey objects"""
+        return RegistryKey(
+            os.path.join(self.path, sub_key),
+            winreg.OpenKeyEx(self.handle, sub_key, access=winreg.KEY_READ),
+        )
+
+    def get_value(self, val_name):
+        """Returns value associated with this key in RegistryValue object"""
+        return RegistryValue(val_name, winreg.QueryValueEx(self.handle, val_name)[0], self.handle)
+
+
+class _HKEY_CONSTANT(RegistryKey):
+    """Subclass of RegistryKey to represent the prebaked, always open registry HKEY constants"""
+
+    def __init__(self, hkey_constant):
+        hkey_name = hkey_constant
+        self._handle = None
+        # because we can't referene winreg yet, handle is none for now
```

Comment:
(minor) might be worth mentioning that is because this class is instantiated at module import time (down in `HKEY`), since it is atypical for that to happen (it took me a minute to realize what you meant).

Reply:
Sounds good, I can see how that would be confusing if you're not writing this module. Also I'll fix the spelling in this docstring.